tests/Unit/Emogrifier/HtmlProcessor/HtmlPrunerTest.php

```diff
@@ -85,4 +85,255 @@ public function removeInvisibleNodesRemovesNodesWithDisplayNone($displayNone)
+
+        self::assertContains($html, $subject->render());
+    }
+
+    /**
+     * @return (string|string[])[][]
+     */
+    public function nonMatchedClassesDataProvider()
+    {
+        return [
+            '1 attribute, 1 class, no classes to keep' => [
+                'html' => '<p class="foo">hello</p>',
+                'classes to keep' => [],
+                'classes expected to be removed' => ['foo'],
+            ],
+            '2 attributes, 1 different class each, no classes to keep' => [
+                'html' => '<p class="foo">hello</p><p class="bar">world</p>',
+                'classes to keep' => [],
+                'classes expected to be removed' => ['foo', 'bar'],
+            ],
+            '1 attribute, 1 class, 1 different class to keep' => [
+                'html' => '<p class="foo">hello</p>',
+                'classes to keep' => ['baz'],
+                'classes expected to be removed' => ['foo'],
+            ],
+            '2 attributes, 1 different class each, 1 different class to keep' => [
+                'html' => '<p class="foo">hello</p><p class="bar">world</p>',
+                'classes to keep' => ['baz'],
+                'classes expected to be removed' => ['foo', 'bar'],
+            ],
+            '2 attributes, same 1 class each, 1 different class to keep' => [
+                'html' => '<p class="foo">hello</p><p class="foo">world</p>',
+                'classes to keep' => ['baz'],
+                'classes expected to be removed' => ['foo'],
+            ],
+            '1 attribute, 2 classes, 1 different class to keep' => [
+                'html' => '<p class="foo bar">hello</p>',
+                'classes to keep' => ['baz'],
+                'classes expected to be removed' => ['foo', 'bar'],
+            ],
+            '1 attribute, 1 class with extra whitespace, 1 different class to keep' => [
+                'html' => '<p class=" foo ">hello</p>',
+                'classes to keep' => ['baz'],
+                'classes expected to be removed' => ['foo'],
+            ],
+            '1 attribute, 2 classes with extra whitespace, 1 different class to keep' => [
+                'html' => '<p class=" foo  bar ">hello</p>',
+                'classes to keep' => ['baz'],
+                'classes expected to be removed' => ['foo', 'bar'],
+            ],
+            '1 attribute, 2 classes separated by newline, 1 different class to keep' => [
+                'html' => "<p class=\"foo\nbar\">hello</p>",
+                'classes to keep' => ['baz'],
+                'classes expected to be removed' => ['foo', 'bar'],
+            ],
+        ];
+    }
+
+    /**
+     * @test
+     *
+     * @param string $html
+     * @param string[] $classesToKeep
+     *
+     * @dataProvider nonMatchedClassesDataProvider
+     */
+    public function removeRedundantClassesRemovesClassAttributesContainingNoClassesToKeep($html, array $classesToKeep)
```

Comment:
I find it confusing that the test has only two parameters, but the data provider provides three. I know that the DP gets recycling in another place, and don't have a good idea to solve this at the moment, though.

Reply:
From memory, if we add an unused parameter, PHPCS (or PHPMD) will complain.